### PR TITLE
Add always fetch latest version

### PR DESCRIPTION
**Pull request by jelly**
---

# Always pull the latest image when creating a new container

When a local image `fedora:latest` is available but a newer version is available on the image registry, checking the "pull latest image" checkbox will pull the latest version before creating the new container

image

### CODE REVIEW

File: test/check-application [hunk 1]
```diff
@@ -1241,6 +1246,39 @@ class TestApplication(testlib.MachineCase):
...
+        # Local registry
+        b.set_input_text("#create-image-image-select-typeahead", "my-busybox")
+        b.click('button.pf-c-toggle-group__button:contains("localhost:5000")')
+        b.wait_text("button.pf-c-select__menu-item:not(.pf-m-disabled)", "localhost:5000/my-busybox")
+
+        # Select image
+        b.click('button.pf-c-select__menu-item:contains("localhost:5000/my-busybox")')
+
+        # Pull the latest image
+        b.click("#run-image-dialog-pull-latest-image")
```
Comment (marusak) [inline test/check-application:1273]: `b.set_checked("#run-image-dialog-pull-latest-image", True)`

File: test/check-application [hunk 2]
```diff
@@ -1769,7 +1807,7 @@ class TestApplication(testlib.MachineCase):
         b.click('.pf-c-modal-box__footer button:contains("Create")')
         self.waitContainerRow(container_name)
 
-    @testlib.skipImage("podman-restart not available in debian/ubuntu", *DISTROS_WITHOUT_PODMAN_RESTART)
+    @ testlib.skipImage("podman-restart not available in debian/ubuntu", *DISTROS_WITHOUT_PODMAN_RESTART)
```
Comment (marusak) [inline test/check-application:1810]: Why these spaces in decorators?

**Review by marusak — CHANGES_REQUESTED** [review]
Do we also test, that when you don't check in the new checkbox then it does not pull?

**Review by marusak — APPROVED** [review]
Thanks!

**Comment by garrett** [comment]
Initial reactions:

- There's not enough vertical space between the checkboxes. Is it using the correct elements? (Not saying space should be manually added... it's just weird that there isn't more. There should be, so this points to incorrect usage of elements... or a bug.)
- Start should be after pull, conceptually.

**Comment by garrett** [comment]
So this is weird on a create dialog when there is no local image. Of course it'd pull the latest image, right? So then it's confusing why it's there.

I am pretty sure when we talked about this that it would probably only exist when there's an existing image and I think we talked about it showing up in the image section, under the image selector, and only when there's a match with a local image.

(Note: Between the holidays and currently being sick including a headache, I could be mistaken. But still, I think it's confusing when there's no local image, so I think this probably still needs work regardless of what we did or didn't talk about in the past.)

**Comment by jelly** [comment]
> 

That's now implemented, not 100% sure if we discussed it before the implementation but it's certainly something which should be in the PR. Thanks for noticing!

**Comment by jelly** [comment]
> So this is weird on a create dialog when there is no local image. Of course it'd pull the latest image, right? So then it's confusing why it's there.
> 
> I am pretty sure when we talked about this that it would probably only exist when there's an existing image and I think we talked about it showing up in the image section, under the image selector, and only when there's a match with a local image.

Moved the pull on create checkbox under the image select, but to avoid it from making the whole form jump it's always shown but only enabled for local images.

**Comment by garrett** [comment]
> Moved the pull on create checkbox under the image select, but to avoid it from making the whole form jump it's always shown but only enabled for local images.

But we know when the image is being used in the image flow and should show it by default and when create container is selected, it's usually not going to be using a local image.

When we have some kind of "cloning" / "rebuild" option, it would also use a local image by default too.

Having a pull option always present even when it doesn't apply would be confusing.

(I'll look at this PR on my computer in a little bit. It's next in the queue now. :wink:)

**Comment by jelly** [comment]
> > Moved the pull on create checkbox under the image select, but to avoid it from making the whole form jump it's always shown but only enabled for local images.
> 
> But we know when the image is being used in the image flow and should show it by default and when create container is selected, it's usually not going to be using a local image.

That's sort of true, if you create an image from a local container from the `Images` dialog sure it shouldn't be shown. But if you create it from the Containers dialog you can select local and remote images there I disabled it.

So, good point we should hide it when you create it from the `Images` section and keep it disabled/enabled for the `Containers` flow.

**Comment by garrett** [comment]
> So, good point we should hide it when you create it from the Images section and keep it disabled/enabled for the Containers flow.

I was suggesting the opposite of that, actually. Show it for the images section (if there's a remote image available... that is, show it if it's not a locally created image) and don't show it by default for the create containers flow, unless it matches a local image in the selector.

**Comment by jelly** [comment]
> > So, good point we should hide it when you create it from the Images section and keep it disabled/enabled for the Containers flow.
> 
> I was suggesting the opposite of that, actually. Show it for the images section (if there's a remote image available... that is, show it if it's not a locally created image) and don't show it by default for the create containers flow, unless it matches a local image in the selector.

Seems my brain mixed things up, sorry. We can do that, do we accept that the dialog will "jump" if you select a local image?

**Comment by garrett** [comment]
> Seems my brain mixed things up, sorry. 

I'm grappling with a headache and feeling icky from being sick still (better, but still ugh)... so I'm just as likely to mix things up or suggest the wrong thing.

> We can do that, do we accept that the dialog will "jump" if you select a local image or can we keep some spacing so it doesn't occur.

We could have a statusline using a helpertext item there in place of a checkbox. ("The image will be downloaded from XYZ." or something like it could work... although it's probably redundant.) The helptertext widget, as it uses small text, might have a different line-height and we might need to adjust it to be the same height to not have a jump. But it'd be better than just a big gap there still... and this is a way to not have a jump.

...Or we just accept that the checkbox might jump into existence in some cases. We do have dialogs that change height already, as the result of interacting widgets above the content that changes.

**Comment by jelly** [comment]
> > Seems my brain mixed things up, sorry.
> 
> I'm grappling with a headache and feeling icky from being sick still (better, but still ugh)... so I'm probably even more likely to mix things up or suggest the wrong thing or not be clear enough. wink
> 
> > We can do that, do we accept that the dialog will "jump" if you select a local image or can we keep some spacing so it doesn't occur.
> We could have a statusline using a helpertext item there in place of a checkbox. ("The image will be downloaded from XYZ." or something like it could work... although it's probably redundant.) The helptertext widget, as it uses small text, might have a different line-height and we might need to adjust it to be the same height to not have a jump. But it'd be better than just a big gap there still... and this is a way to not have a jump.

I think that adds to much redundant information, something which we should avoid.

> ...Or we just accept that the checkbox might jump into existence in some cases. We do have dialogs that change height already, as the result of interacting widgets above the content that changes.

That's the best compromise in my opinion, the new updated PR now shows it by default for the `Images` overview and it is shown when required for the "Create container" workflow under "Containers", the jumping seems acceptable and maybe something we can improve on later with new ideas.

**Comment by jelly** [comment]
> Here's what the autocomplete is covering for me:
> 
> image
> 
> Additionally (out of scope of this PR): What is the `<none:none>`?

I have <none:none> too hmm, not sure where it comes from but it's also shown in `podman images`.

**Comment by marusak** [comment]
`none:none` are intermediate images. We show them in the listing, but there is button on the bottom to show them.

**Review by marusak — CHANGES_REQUESTED** [review]
Looks good, thanks!
Can you please add test that we don't show this checkbox when pulling the image?

**Comment by jelly** [comment]
> Looks good, thanks! Can you please add test that we don't show this checkbox when pulling the image?

added!

File: test/check-application [hunk 3]
```diff
@@ -1241,10 +1246,46 @@ class TestApplication(testlib.MachineCase):
...
+
+        # Local registry
+        b.set_input_text("#create-image-image-select-typeahead", "my-busybox")
+        b.click('button.pf-c-toggle-group__button:contains("Local")')
+
+        # Select image
+        b.click('button.pf-c-select__menu-item:contains("localhost:5000/my-busybox")')
+
+        # Pull the latest image
+        # testlib.sit()
```
Comment (marusak) [inline test/check-application:1279]: :)

**Review by marusak — APPROVED** [review]
Awesome, thank you!
Can you please update screenshot as well for the release notes?

**Comment by jelly** [comment]
> Awesome, thank you! Can you please update screenshot as well for the release notes?

done!

**Pull request merged**